src/main/ruby/truffleruby/core/dir_glob.rb

```diff
@@ -245,13 +245,15 @@ def process_directory(matches, parent, entry, glob_base_dir)
             type = entry.type
 
             full = path_join(path, ent)
-            if (type == Truffle::DirOperations::DT_DIR) and (allow_dots or ent.getbyte(0) != 46) # ?.
-              @next.process_entry ent, type, matches, path, glob_base_dir
-
-              stack << [path, dir_entries]
-              path = full
-              dir_entries = subdir_entries(glob_base_dir, full)
-            elsif (allow_dots or ent.getbyte(0) != 46) # ?.
+            if (type == Truffle::DirOperations::DT_DIR)
+              if (allow_dots or ent.getbyte(0) != 46) # ?.
+                @next.process_entry ent, type, matches, path, glob_base_dir
+
+                stack << [path, dir_entries]
+                path = full
+                dir_entries = subdir_entries(glob_base_dir, full)
+              end
+            else
```

Comment:
i removed the `allowed or not a dot` condition for files as i think the `match?` called by process_entry handles that check.

Reply:
It's definitely possible that we need more specs to help get this right.

Without this change i get this failure:
```
      Dir.glob('**/.*', File::FNM_DOTMATCH, base: "deeply/nested").sort.should == expected

Expected [] == [".dotfile.ext", "directory/structure/.ext"]
to be truthy but was false
```

I changed this because it used to say
```
if this is a dir and (allow_dots or starts with non-dot)
  process this as a dir
elsif allow_dots or starts-with-non-dot
  process this as a file
end
```
in the scenario of the failing test case above:
- item is a dir
- name starts with a dot
- allow_dots is false (no FNM_MATCH) *but* the glob matches dots

so this fails both of the old conditions and doesn't get processed at all.

So I think the first condition needs to be "if a dir" separate from the "dot check"
so that we don't accidentally process dirs as files.

Then we need to figure out the right condition for "should entry be processed based on its name". 

i think i was looking at https://github.com/oracle/truffleruby/blob/bd29887e08f7a5ab52ec42349cd8ecc5a8bc1785/src/main/ruby/truffleruby/core/dir_glob.rb#L302-L304
and seeing that the `process_entry` method was doing the `match?(entry)` check
but you are right, i do see that there are some other `process_entry` definitions that don't have that check.

I also see that this same `if` branch is used in another class in this file:
https://github.com/oracle/truffleruby/blob/bd29887e08f7a5ab52ec42349cd8ecc5a8bc1785/src/main/ruby/truffleruby/core/dir_glob.rb#L204-L214

I probably just need to step back and read through the whole file.
... and add some more specs.

Reply:
also i think at this point in the code there isn't access to the glob itself (for us to be able to enhance the check),
but the glob is built in to the `@next` item, since that where the `match?` is happening.

so this might be the structure that we need if we just verify that all the resulting process_entry methods do the `match?` check that they should.

Reply:
> Some nodes don't check File::FNM_DOTMATCH flag in process_entry e.g. EntryMatch and ConstantEntry

`EntryMatch` does check it:
https://github.com/oracle/truffleruby/blob/797ac69479f54360835ec38e5da942386563bd4a/src/main/ruby/truffleruby/core/dir_glob.rb#L319-L321
https://github.com/oracle/truffleruby/blob/797ac69479f54360835ec38e5da942386563bd4a/src/main/ruby/truffleruby/core/dir_glob.rb#L274-L276

`ConstantEntry` does not, but i believe that is the desired behavior, isn't it?
It is used when the last path part is a string without a glob (which means FNM_DOTMATCH would not affect the match) and the method checks the name directly.
https://github.com/oracle/truffleruby/blob/797ac69479f54360835ec38e5da942386563bd4a/src/main/ruby/truffleruby/core/dir_glob.rb#L162-L166


adding a spec for this case, the old code fails and the new code passes.
```
    Dir.glob('**/.dotfile', base: ".").sort.should == expected
Expected [] == [".dotfile", "deeply/.dotfile", "subdir_one/.dotfile"]
to be truthy but was false
```
```

Reply:
i have several more specs i added locally to cover additional scenarios (like the ConstantEntry), but i'll wait to see if we want to change anything else before i push them

Reply:
I see. Great! So it's actually a fix, not a refactoring as it seemed. 

It was surprising to me that expected behaviour is that a pattern with explicit leading `.` matches hidden files/directories by default (without `File::FNM_DOTMATCH` flag).